pkg/controllers/nodetemplate/infrastructure.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/aws/karpenter/pkg/apis/v1alpha1"
+	"github.com/aws/karpenter/pkg/controllers/providers"
+)
+
+type Infrastructure struct {
+	kubeClient          client.Client
+	sqsProvider         *providers.SQS
+	eventBridgeProvider *providers.EventBridge
+
+	lastInfrastructureReconcile time.Time // Keeps track of the last reconcile time for infra, so we don't keep calling APIs
+}
+
+// Reconcile reconciles the infrastructure based on whether interruption handling is enabled and deletes
+// the infrastructure by ref-counting when the last AWSNodeTemplate is removed
+func (i *Infrastructure) Reconcile(ctx context.Context, nodeTemplate *v1alpha1.AWSNodeTemplate) (reconcile.Result, error) {
+	if awssettings.FromContext(ctx).EnableInterruptionHandling {
+		list := &v1alpha1.AWSNodeTemplateList{}
+		if err := i.kubeClient.List(ctx, list); err != nil {
+			return reconcile.Result{}, err
+		}
+		if !nodeTemplate.DeletionTimestamp.IsZero() && len(list.Items) == 1 {
+			if err := i.DeleteInfrastructure(ctx); err != nil {
+				return reconcile.Result{}, err
+			}
+			i.lastInfrastructureReconcile = time.Time{}
+			return reconcile.Result{}, nil
+		} else if len(list.Items) >= 1 {
+			if i.lastInfrastructureReconcile.Add(time.Hour).Before(time.Now()) {
+				if err := i.CreateInfrastructure(ctx); err != nil {
+					infrastructureHealthy.Set(0)
+					return reconcile.Result{}, err
+				}
+				i.lastInfrastructureReconcile = time.Now()
+				infrastructureHealthy.Set(1)
+			}
+		}
+	} else {
+		infrastructureHealthy.Set(0)
+	}
+
+	// TODO: Implement an alerting mechanism for settings updates; until then, just poll
+	return reconcile.Result{RequeueAfter: time.Second * 10}, nil
+}
+
+// CreateInfrastructure provisions an SQS queue and EventBridge rules to enable interruption handling
+func (i *Infrastructure) CreateInfrastructure(ctx context.Context) error {
+	defer metrics.Measure(infrastructureCreateDuration)()
+	if err := i.ensureQueue(ctx); err != nil {
+		return fmt.Errorf("ensuring queue, %w", err)
+	}
+	if err := i.ensureEventBridge(ctx); err != nil {
+		return fmt.Errorf("ensuring eventBridge rules and targets, %w", err)
+	}
+	logging.FromContext(ctx).Infof("Completed reconciliation of infrastructure")
+	return nil
+}
+
+// DeleteInfrastructure removes the infrastructure that was stood up and reconciled
+// by the infrastructure controller for SQS message polling
+func (i *Infrastructure) DeleteInfrastructure(ctx context.Context) error {
+	defer metrics.Measure(infrastructureDeleteDuration)()
+	logging.FromContext(ctx).Infof("Deprovisioning the infrastructure...")
+	funcs := []func(context.Context) error{
+		i.deleteQueue,
+		i.deleteEventBridge,
+	}
+	errs := make([]error, len(funcs))
+	workqueue.ParallelizeUntil(ctx, len(funcs), len(funcs), func(i int) {
+		errs[i] = funcs[i](ctx)
+	})
+
+	err := multierr.Combine(errs...)
+	if err != nil {
+		return err
+	}
+	logging.FromContext(ctx).Infof("Completed deprovisioning the infrastructure")
```

Comment:
I worry users might take the current message out of context.
```suggestion
	logging.FromContext(ctx).Infof("Ensured cleanup of EC2 interruption handling infrastructure")
```